### ISSUE TITLE
Warn when mutably borrowing an unsafe block that returnes a derefenced Copy value

### ISSUE THREAD

**Issue by DHauckCW** [rust-lang/rust-clippy#15893]
### What it does

Code of the form

```rust
let x = &mut unsafe { *y };
```

where the type of `y` is `Copy` successfully compiles, but very likely does not correspond to the intended behaviour.

### Advantage

Avoid accidental copying of a value, which leads to the original object being unchanged.

### Drawbacks

The code might be written as intended, especially if there are multiple statements inside the unsafe block.

### Example

```rust
let mut x: u32 = 4; // Somewhere
// ...
let x_ptr: *mut u32 = &raw mut x;
let y = &mut unsafe { *x_ptr };
```

Could be meant to be written as:

```rust
let y = unsafe { &mut *x_ptr };
```

or

```rust
let y = unsafe { x_ptr.as_mut().unwrap() }
```

or with nightly

```rust
let y = unsafe { x_ptr.as_mut_unchecked() }
```

### Additional Context

The actual behaviour is in my opinion better written as

```rust
let y = unsafe { &mut (*x_ptr).clone() };
```

or

```rust
let mut z = unsafe { *x_ptr };
let x = &mut z;
```

to make the copy more explicit.

**Comment by y21** [comment]
#13984 implements basically this (slightly more general, it lints regardless of whether it's an unsafe block or a normal block).

**Comment by DHauckCW** [comment]
Great! Then my search skills were not strong enough.
